### ISSUE TITLE
示例程序报错

### ISSUE THREAD

**Issue by retroactively** [SupRL/WorldModelTrainer#1]
根据readme配置好环境后:

`cd src`
`python3 train.py`

![iShot_2023-08-14_20 39 08](https://github.com/SupRL/WorldModelTrainer/assets/41630114/eeedf9b2-041f-492e-862f-8dad528b0c45)

**Comment by SupRL** [comment]
操作系统和Python版本以及Python包的版本可以给一下么？

**Comment by retroactively** [comment]
OS: Macos13.4 (Intel)
Python: 3.8.17
其他包是根据readme和requirements.txt下载的。

Name                    Version                   Build  Channel
accelerate                0.19.0                   pypi_0    pypi
aiozmq                    1.0.0                    pypi_0    pypi
appdirs                   1.4.4                    pypi_0    pypi
bzip2                     1.0.8                h0d85af4_4    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
ca-certificates           2023.7.22            h8857fd0_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
certifi                   2023.7.22                pypi_0    pypi
charset-normalizer        3.2.0                    pypi_0    pypi
click                     8.1.6                    pypi_0    pypi
cloudpickle               2.2.1                    pypi_0    pypi
contourpy                 1.1.0                    pypi_0    pypi
cycler                    0.11.0                   pypi_0    pypi
docker-pycreds            0.4.0                    pypi_0    pypi
filelock                  3.12.2                   pypi_0    pypi
flatten-dict              0.4.2                    pypi_0    pypi
fonttools                 4.42.0                   pypi_0    pypi
fsspec                    2023.6.0                 pypi_0    pypi
gitdb                     4.0.10                   pypi_0    pypi
gitpython                 3.1.32                   pypi_0    pypi
gym                       0.26.2                   pypi_0    pypi
gym-notices               0.0.8                    pypi_0    pypi
huggingface-hub           0.16.4                   pypi_0    pypi
idna                      3.4                      pypi_0    pypi
importlib-metadata        6.8.0                    pypi_0    pypi
importlib-resources       6.0.1                    pypi_0    pypi
kiwisolver                1.4.4                    pypi_0    pypi
libffi                    3.4.2                h0d85af4_5    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libsqlite                 3.42.0               h58db7d2_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libzlib                   1.2.13               h8a1eda9_5    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
matplotlib                3.7.1                    pypi_0    pypi
msgpack                   1.0.5                    pypi_0    pypi
msgpack-numpy             0.4.8                    pypi_0    pypi
ncurses                   6.4                  hf0c8a7f_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
numpy                     1.24.4                   pypi_0    pypi
openssl                   3.1.2                h8a1eda9_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
packaging                 23.1                     pypi_0    pypi
pandas                    1.5.3                    pypi_0    pypi
pathtools                 0.1.2                    pypi_0    pypi
pillow                    10.0.0                   pypi_0    pypi
pip                       23.2.1             pyhd8ed1ab_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
protobuf                  3.20.3                   pypi_0    pypi
psutil                    5.9.5                    pypi_0    pypi
pyparsing                 3.1.1                    pypi_0    pypi
python                    3.8.17          hf9b03c3_0_cpython    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
python-dateutil           2.8.2                    pypi_0    pypi
pytypes                   1.0b10                   pypi_0    pypi
pytz                      2023.3                   pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
pyzmq                     25.1.1                   pypi_0    pypi
readline                  8.2                  h9e318b2_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
regex                     2023.8.8                 pypi_0    pypi
requests                  2.31.0                   pypi_0    pypi
sentry-sdk                1.29.2                   pypi_0    pypi
setproctitle              1.3.2                    pypi_0    pypi
setuptools                59.5.0                   pypi_0    pypi
six                       1.16.0                   pypi_0    pypi
smmap                     5.0.0                    pypi_0    pypi
tensorboardx              2.6                      pypi_0    pypi
tk                        8.6.12               h5dbffcc_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
tokenizers                0.13.3                   pypi_0    pypi
torch                     1.9.1                    pypi_0    pypi
torchaudio                0.9.1                    pypi_0    pypi
torchvision               0.10.1                   pypi_0    pypi
tqdm                      4.66.1                   pypi_0    pypi
transformers              4.29.0                   pypi_0    pypi
transforms3d              0.4.1                    pypi_0    pypi
typing-extensions         4.7.1                    pypi_0    pypi
urllib3                   2.0.4                    pypi_0    pypi
wandb                     0.13.10                  pypi_0    pypi
wheel                     0.41.1             pyhd8ed1ab_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
xz                        5.2.6                h775f41a_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
zipp                      3.16.2                   pypi_0    pypi
zmq                       0.0.0                    pypi_0    pypi

**Comment by retroactively** [comment]
其他人在windwos平台下跑的时候，也有这个问题的。

**Comment by SupRL** [comment]
您好，windows和macos确实可能会有问题，看这个报错应该主要是由于pytorch的dataloader开启多进程加载时操作系统层面的一些实现不同导致的报错，如果一定要用windows和macos的话，可以尝试将https://github.com/SupRL/WorldModelTrainer/blob/fdf6a46bdaf4f7d186b9e6fd4f7a5d08369da13e/src/train.py#L241 和https://github.com/SupRL/WorldModelTrainer/blob/fdf6a46bdaf4f7d186b9e6fd4f7a5d08369da13e/src/train.py#L296 两个地方的num_workers和test_num_workers都设置为0试一下，但是这样可能会拖慢训练的速度，另外还是建议用ubuntu20.04操作系统进行训练，macos和windows我们没有进行过系统的测试，稍后会在readme中补充操作系统的要求，感谢反馈。

**Comment by retroactively** [comment]
您好，上面给出的解决方案还是会报错：
1. num_workers 必须要 > 0;  (简单改成正数也不起作用的)
2. 代码还未跑到 test_num_workers 这里， test_num_workers 改成 0 也没用，到251行已经报错结束了（和前面的报错一样）；

**Comment by SupRL** [comment]
您好，可以做如下修改：
1. 将下面这个地方的num_workers设置为0
https://github.com/SupRL/WorldModelTrainer/blob/fdf6a46bdaf4f7d186b9e6fd4f7a5d08369da13e/src/train.py#L241
2. 将下面这一行中的prefetch_factor=5去掉
https://github.com/SupRL/WorldModelTrainer/blob/fdf6a46bdaf4f7d186b9e6fd4f7a5d08369da13e/src/train.py#L243
即修改为
`train_dataloader = torch.utils.data.DataLoader(train_dataset, batch_size=math.ceil(max_data_num/batch_size) * batch_size,
                                                           pin_memory=True, drop_last=True, num_workers=num_workers)`
3.test部分的test_num_workers和test_dataloader进行类似的修改
